### PR TITLE
Fdmapping random access

### PR DESCRIPTION
**Pull request by hbrodin**

### CODE REVIEW

File: polytracker/taint_dag.py [hunk 1]
```diff
@@ -117,17 +118,15 @@ def __init__(self, buffer: bytearray) -> None:
...
-        end = offset + self.header.fd_mapping_size
-
-        while offset < end:
-            fdhdr = TDFDHeader.from_buffer_copy(self.buffer, offset)
-            offset += sizeof(TDFDHeader)
-            path = str(self.buffer[offset : offset + fdhdr.namelen], "utf-8")
+        for i in range(0, self.header.fd_mapping_count):
+            header_offset = offset + sizeof(TDFDHeader)*i
+            fdhdr = TDFDHeader.from_buffer_copy(self.buffer, header_offset)
+            sbegin = offset + fdhdr.name_offset
```
Comment (surovic) [inline polytracker/taint_dag.py:127]: Am I correct in saying that the file name offset is relative to the start of the fd header section? Given the computation above.

**Review by surovic — APPROVED** [review]

File: polytracker/include/taintdag/fdmapping.hpp [hunk 2]
```diff
@@ -10,121 +10,163 @@
...
+        if (n == 0)
+          return {};
 
-        if (found) {
-          std::optional<taint_range_t> r;
-          if (found->prealloc_begin != 0)
-            r = taint_range_t{found->prealloc_begin, found->prealloc_end};
-          return std::make_pair(found_idx, r);
+        auto latesthdr = &get(n-1);
+        // NOTE (hbrodin): Assumes first is not mapped in the first sizeof(FDMappingHdr) bytes of the address space.
```
[the remark this answers is not in the document]
Comment (hbrodin) [reply polytracker/include/taintdag/fdmapping.hpp:131]: No, it doesn't. I'll add some more comments to describe the iteration.

File: polytracker/dumptdag.py [hunk 3]
```diff
@@ -103,18 +104,21 @@ def __init__(self, mm: bytearray) -> None:
         self.hdr = FileHdr.from_buffer_copy(mm)
         self.mm = mm
 
-    def fd_mappings(self) -> Iterable[Tuple[str, int, int]]:
-        offset = self.hdr.fd_mapping_offset
-        end = offset + self.hdr.fd_mapping_size
+    def fd_mapping(self, index: int) -> Union[None, Tuple[str, int, int]]:
```
[the remark this answers is not in the document]
Comment (hbrodin) [reply polytracker/dumptdag.py:107]: Good suggestion. I'll fix.

File: polytracker/taint_dag.py (hunk 1 above)
Comment (hbrodin) [reply polytracker/taint_dag.py:127]: Yes. See also the comment in fdmapping.hpp.

**Pull request merged**